exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unnest/UnnestRecordBatch.java

```diff
@@ -0,0 +1,451 @@
+    // updated the incoming vector.
+    // We do, however, need to call doWork() to do the actual work.
+    // We also need to handle schema build if it is the first batch
+
+    if ((state == BatchState.FIRST)) {
+      state = BatchState.NOT_FIRST;
+      try {
+        stats.startSetup();
+        hasRemainder = true; // next call to next will handle the actual data.
+        logger.debug("First batch received");
+        schemaChanged(); // checks if schema has changed (redundant in this case becaause it has) AND saves the
+                         // current field metadata for check in subsequent iterations
+        setupNewSchema();
+      } catch (SchemaChangeException ex) {
+        kill(false);
+        logger.error("Failure during query", ex);
+        context.getExecutorState().fail(ex);
+        return IterOutcome.STOP;
+      } finally {
+        stats.stopSetup();
+      }
+      return IterOutcome.OK_NEW_SCHEMA;
+    } else {
+      assert state != BatchState.FIRST : "First batch should be OK_NEW_SCHEMA";
+      container.zeroVectors();
+
+      // Check if schema has changed
+      if (lateral.getRecordIndex() == 0 && schemaChanged()) {
+        hasRemainder = true;     // next call to next will handle the actual data.
+        try {
+          setupNewSchema();
+        } catch (SchemaChangeException ex) {
+          kill(false);
+          logger.error("Failure during query", ex);
+          context.getExecutorState().fail(ex);
+          return IterOutcome.STOP;
+        }
+        return OK_NEW_SCHEMA;
+      }
+      if (lateral.getRecordIndex() == 0) {
+        unnest.resetGroupIndex();
+      }
+      return doWork();
+    }
+
+  }
+
+    @Override
+  public VectorContainer getOutgoingContainer() {
+    return this.container;
+  }
+
+  @SuppressWarnings("resource") private void setUnnestVector() {
+    final TypedFieldId typedFieldId = incoming.getValueVectorId(popConfig.getColumn());
+    final MaterializedField field = incoming.getSchema().getColumn(typedFieldId.getFieldIds()[0]);
+    final RepeatedValueVector vector;
+    final ValueVector inVV =
+        incoming.getValueAccessorById(field.getValueClass(), typedFieldId.getFieldIds()).getValueVector();
+
+    if (!(inVV instanceof RepeatedValueVector)) {
+      if (incoming.getRecordCount() != 0) {
+        throw UserException.unsupportedError().message("Unnest does not support inputs of non-list values.")
+            .build(logger);
+      }
+      // Inherited from FLATTEN. When does this happen???
+      //when incoming recordCount is 0, don't throw exception since the type being seen here is not solid
+      logger.error("setUnnestVector cast failed and recordcount is 0, create empty vector anyway.");
+      vector = new RepeatedMapVector(field, oContext.getAllocator(), null);
+    } else {
+      vector = RepeatedValueVector.class.cast(inVV);
+    }
+    unnest.setUnnestField(vector);
+  }
+
+  @Override
+  protected IterOutcome doWork() {
```

Comment:
Flatten's doWork() could potentially throw an OOM.  Can a similar situation occur for Unnest ?  (although much less likely due to the row-by-row nature of unnest when running in the context of Lateral join).